fix: revert docarray explicit transitive dependencies and use extras instead

malandis opened the pull request:
tldr: The docarray [integration
PR](https://github.com/hwchase17/langchain/pull/4483) introduced a pinned dependency to protobuf. This is a docarray dependency, not a langchain dependency. Since this is handled by the docarray dependencies, it is unnecessary here.

Further, as a pinned dependency, this quickly leads to incompatibilities with application code that consumes the library. Much less with a heavily used library like protobuf.

Detail: as we see in the [docarray
integration](https://github.com/hwchase17/langchain/pull/4483/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R81-R83), the transitive dependencies of docarray were also listed as langchain dependencies. This is unnecessary as the docarray project has an appropriate [extras](https://github.com/docarray/docarray/blob/a01a05542d17264b8a164bec783633658deeedb8/pyproject.toml#L70). The docarray project also does not require this _pinned_ version of protobuf, rather [a minimum version](https://github.com/docarray/docarray/blob/a01a05542d17264b8a164bec783633658deeedb8/pyproject.toml#L41). So this pinned version was likely in error.

To fix this, this PR reverts the explicit hnswlib and protobuf dependencies and adds the hnswlib extras install for docarray (which installs hnswlib and protobuf, as originally intended). Because version `0.32.0`
of the docarray hnswlib extras added protobuf, we bump the docarray dependency from `^0.31.0` to `^0.32.0`.

# revert docarray explicit transitive dependencies and use extras instead

## Who can review?

@dev2049 -- reviewed the original PR
@eyurtsev -- bumped the pinned protobuf dependency a few days ago

malandis committed:
fix: use docarray extras instead of specifying separate deps

The docarray [integration
PR](https://github.com/hwchase17/langchain/pull/4483) introduced a
pinned dependency to protobuf.

As library developers, we should avoid pinned dependencies as this
quickly leads to incompatibilities with application code. Much less
with a heavily used library like protobuf.

As we see in the [docarray
integration](https://github.com/hwchase17/langchain/pull/4483/files#diff-50c86b7ed8ac2cf95bd48334961bf0530cdc77b5a56f852c5c61b89d735fd711R81-R83),
the transitive dependencies of docarray were also listed as langchain
dependencies. This is unnecessary as the docarray project has an
[extras
install](https://github.com/docarray/docarray/blob/a01a05542d17264b8a164bec783633658deeedb8/pyproject.toml#L70)
that lists these.

This PR reverts the explicit hnswlib and protobuf dependencies and
adds the hnswlib extras install for docarray (which installs hnswlib
and protobuf). Because version 0.32.0
of the docarray hnswlib extras added protobuf, we bump the docarray
dependency from `^0.31.0` to `^0.32.0`.

malandis committed:
Merge branch 'master' into fix/remove-protobuf-dependency

Also fix merge conflicts in pyproject.toml and poetry.lock.

hwchase17 commented on pyproject.toml:
```diff
@@ -181,10 +179,10 @@ openai = ["openai", "tiktoken"]
 cohere = ["cohere"]
 in_memory_store = ["docarray"]
-hnswlib = ["docarray", "protobuf", "hnswlib"]
+hnswlib = ["docarray"]
```
having `hnswlib` have the docarray dependency only seems weird. should call `docarray`?

malandis replied:
I thought that was strange too but had not opined here. I agree and will change.

malandis replied:
Added a "docarray" extra in https://github.com/hwchase17/langchain/pull/5015/commits/07a4ded643441e021e0c2634f1d60298f699f2b2

malandis committed:
chore: clarify docarray extras name

An extra was previously introduced for "hnswlib" which only installed
"docarray". While "hnswlib" is a dependency of docarray, docarray is a
separate integration, so this is misleading to users.

There was also a separate extras for "in_memory_store". This is again
misleading since there are various in memory stores in the system that
are very specific.

hwchase17 reviewed (approved):
lgtm!

dev2049 committed:
merge

eyurtsev commented:
This sounds good. I was trying to provide a minimum protobuf constraint before, but poetry locking was failing. If removing it entirely is working then great!
